File: frontend/src/api/leads.ts
import { apiFetch, parseErrorDetail } from "./client";

export interface Lead {
  id: string;
  phone: string;
  name: string;
  lang: string;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface LeadDetail extends Lead {
  call_count: number;
  last_call_at: string | null;
}

export interface PaginatedLeads {
  items: Lead[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateLeadInput {
  phone: string;
  name: string;
  lang: string;
}

export interface UpdateLeadInput {
  phone?: string;
  name?: string;
  lang?: string;
  status?: string;
}

export interface FetchLeadsParams {
  page?: number;
  pageSize?: number;
  q?: string;
  status?: string;
  lang?: string;
}

export async function fetchLeads(params: FetchLeadsParams = {}): Promise<PaginatedLeads> {
  const search = new URLSearchParams();
  search.set("page", String(params.page ?? 1));
  search.set("page_size", String(params.pageSize ?? 20));
  if (params.q) search.set("q", params.q);
  if (params.status) search.set("status", params.status);
  if (params.lang) search.set("lang", params.lang);

  const response = await apiFetch(`/leads?${search.toString()}`);
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, `Failed to fetch leads: ${response.status}`));
  }
  return response.json();
}

export async function fetchLead(id: string): Promise<LeadDetail> {
  const response = await apiFetch(`/leads/${id}`);
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, `Failed to fetch lead: ${response.status}`));
  }
  return response.json();
}

export async function createLead(input: CreateLeadInput): Promise<Lead> {
  const response = await apiFetch("/leads", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, `Failed to create lead: ${response.status}`));
  }
  return response.json();
}

export async function updateLead(id: string, patch: UpdateLeadInput): Promise<Lead> {
  const response = await apiFetch(`/leads/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(patch),
  });
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response, `Failed to update lead: ${response.status}`));
  }
  return response.json();
}

export async function deleteLead(id: string): Promise<void> {
  const response = await apiFetch(`/leads/${id}`, { method: "DELETE" });
  if (!response.ok && response.status !== 204) {
    throw new Error(await parseErrorDetail(response, `Failed to delete lead: ${response.status}`));
  }
}

File: backend/app/routers/leads.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session

from app.core.clock import utcnow
from app.core.config import DEFAULT_ORG_ID
from app.db.models import CallRecord, Lead
from app.db.session import get_db
from app.schemas.leads import (
    CreateLeadRequest,
    LeadDetailResponse,
    LeadResponse,
    PaginatedLeads,
    UpdateLeadRequest,
)

router = APIRouter(prefix="/leads", tags=["leads"])


def _active_leads(db: Session):
    """Base query for leads that have not been soft-deleted. Reuse everywhere."""
    return select(Lead).where(Lead.deleted_at.is_(None))


def _get_active_lead(db: Session, lead_id: uuid.UUID) -> Lead:
    lead = db.scalar(_active_leads(db).where(Lead.id == lead_id))
    if lead is None:
        raise HTTPException(status_code=404, detail="lead not found")
    return lead


@router.post("", response_model=LeadResponse, status_code=201)
def create_lead(body: CreateLeadRequest, db: Session = Depends(get_db)) -> Lead:
    lead = Lead(
        organization_id=DEFAULT_ORG_ID,
        phone=body.phone,
        name=body.name,
        lang=body.lang,
        status="new",
    )
    db.add(lead)
    db.commit()
    db.refresh(lead)
    return lead


@router.get("", response_model=PaginatedLeads)
def list_leads(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: str | None = None,
    status: str | None = None,
    lang: str | None = None,
    db: Session = Depends(get_db),
) -> PaginatedLeads:
    query = _active_leads(db)

    if q:
        query = query.where(or_(Lead.name.ilike(f"%{q}%"), Lead.phone.ilike(f"%{q}%")))
    if status:
        query = query.where(Lead.status == status)
    if lang:
        query = query.where(Lead.lang == lang)

    total = db.scalar(select(func.count()).select_from(query.subquery()))

    items = db.scalars(
        query.order_by(Lead.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
    ).all()

    return PaginatedLeads(items=list(items), total=total or 0, page=page, page_size=page_size)


@router.get("/{lead_id}", response_model=LeadDetailResponse)
def get_lead(lead_id: uuid.UUID, db: Session = Depends(get_db)) -> LeadDetailResponse:
    lead = _get_active_lead(db, lead_id)

    call_count, last_call_at = db.execute(
        select(func.count(CallRecord.id), func.max(CallRecord.started_at)).where(
            CallRecord.lead_id == lead_id
        )
    ).one()

    return LeadDetailResponse(
        id=lead.id,
        phone=lead.phone,
        name=lead.name,
        lang=lead.lang,
        status=lead.status,
        created_at=lead.created_at,
        updated_at=lead.updated_at,
        call_count=call_count or 0,
        last_call_at=last_call_at,
    )


@router.patch("/{lead_id}", response_model=LeadResponse)
def update_lead(lead_id: uuid.UUID, body: UpdateLeadRequest, db: Session = Depends(get_db)) -> Lead:
    lead = _get_active_lead(db, lead_id)

    updates = body.model_dump(exclude_unset=True)
    for field, value in updates.items():
        setattr(lead, field, value)

    if updates:
        lead.updated_at = utcnow()

    db.commit()
    db.refresh(lead)
    return lead


@router.delete("/{lead_id}", status_code=204)
def delete_lead(lead_id: uuid.UUID, db: Session = Depends(get_db)) -> None:
    lead = _get_active_lead(db, lead_id)
    lead.deleted_at = utcnow()
    db.commit()
